modelling/src/main/java/org/axonframework/modelling/command/annotation/InjectEntityParameterResolverFactory.java

```diff
@@ -0,0 +1,112 @@
+ *
+ * @author Mitchell Herrijgers
+ * @see InjectEntity
+ * @since 5.0.0
+ */
+public class InjectEntityParameterResolverFactory implements ParameterResolverFactory {
+
+    private final StateManager stateManager;
+
+    /**
+     * Initialize the factory with the given {@code stateManager} to load entities from.
+     *
+     * @param stateManager The registry to load entities from.
+     */
+    public InjectEntityParameterResolverFactory(@Nonnull StateManager stateManager) {
+        this.stateManager = requireNonNull(stateManager, "The StateManager is required");
+    }
+
+    @Override
+    public ParameterResolver<?> createInstance(Executable executable, Parameter[] parameters, int parameterIndex) {
+        Parameter parameter = parameters[parameterIndex];
+        if (!parameter.isAnnotationPresent(InjectEntity.class)) {
+            return null;
+        }
+
+        InjectEntity annotation = parameter.getAnnotation(InjectEntity.class);
+        EntityIdResolver<?> entityIdResolver = getEntityIdResolver(annotation);
+
+        // If the parameter is a ManagedEntity, we need to extract the actual entity type
+        Class<?> entityClass = parameter.getType();
+        boolean isManagedEntity = ManagedEntity.class.isAssignableFrom(entityClass);
+        if (isManagedEntity) {
+            ParameterizedType parameterizedType = (ParameterizedType) parameter.getParameterizedType();
+            entityClass = (Class<?>) parameterizedType.getActualTypeArguments()[1];
+        }
+        assertStateManagerCanLoadEntityClass(entityClass);
+        return new InjectEntityParameterResolver(
+                stateManager,
+                entityClass,
+                entityIdResolver,
+                isManagedEntity
+        );
+    }
+
+
+    /**
+     * Let's do a boot-time check to see if it's actually valid. As the id type is defined runtime, we can only
+     * check for the entity type. Still better than nothing!
+     *
+     * @param entityClass The entity class to check
+     */
+    private void assertStateManagerCanLoadEntityClass(Class<?> entityClass) {
+        if (stateManager.registeredEntities().stream().noneMatch(entityClass::isAssignableFrom)) {
+            throw new IllegalArgumentException(
+                    "The entity class " + entityClass.getName()
+                            + " is not registered with the StateManager. Can not construct a ParameterResolver for the @InjectModel of said class.");
+        }
+    }
+
+    private static EntityIdResolver<?> getEntityIdResolver(InjectEntity annotation) {
+        if (annotation.idProperty() != null && !annotation.idProperty().isEmpty()) {
+            return new PropertyBasedEntityIdResolver(annotation.idProperty());
+        }
+        return constructCustomEntityIdResolver(annotation);
+    }
+
+    private static EntityIdResolver<?> constructCustomEntityIdResolver(InjectEntity annotation) {
```

Comment:
We are, strictly speaking, not always generating a custom `EntityIdResolver`, right? In absence of both, the `AnnotationBasedEntityIdResolver` will kick in. If I follow the implementation correctly, of course.

Reply:
You are correct. Renaming method to `constructEntityIdResolver`, as it can be the annotation-based one